Fixes #37653 - Always load local disk's GRUB2 configuration

goarsna opened the pull request:
We propose to switch to loading the local disk's GRUB2 configuration regardless of weather SecureBoot is enabled or not. This standardizes the boot process under UEFI.
 
To support SecureBoot for arbitrary operating systems, support for distribution vendor specific boot files will be added by https://github.com/theforeman/foreman/pull/9864 and the related PRs.

Distribution vendors patch GRUB2 differently which results in
 
- different behavior when using `exit 1` (see: https://github.com/theforeman/foreman/pull/10207) and
- differences in the support of the `connectefi` command (available on EL only), which affects chainloading from local disks using the `chainloader` command (see: https://github.com/theforeman/foreman/pull/10126).
 
Therefore, in case distribution vendor specific boot files are set up for the operating system of a host, using the `chainloader` command may lead to problems during boot in case the distribution vendor specific GRUB2 doesn't support the `connectefi` command - which is at least true for non EL systems.
 
For SecureBoot enabled hosts this has already been fixed by switching to loading the local disk's GRUB2 configuration (https://github.com/theforeman/foreman/pull/10207). Now we want to introduce this also for the case that SecureBoot is disabled.
 
For Windows we still use chainloading as there is no local `grub.cfg`. The default EL GRUB2 which is used in this case supports the `connectefi` command. SecureBoot verification is done by the Microsoft certificate in the db.

jloeser commented on app/views/unattended/provisioning_templates/snippet/pxegrub2_chainload.erb:
```diff
@@ -4,26 +4,12 @@ name: pxegrub2_chainload
 snippet: true
 description: |
-  In Foreman's typical PXE workflow, managed hosts are configured to always boot from network and inventory build flag dictates if they should boot into installer (build is on) or boot from local drive (build is off). This template is used to chainload from EFI ESP for systems which booted from network. It is not as straightforward as in BIOS and EFI boot file must be found on an ESP partition.
+  In Foreman's typical PXE workflow, managed hosts are configured to always boot from network and inventory build flag dictates if they should boot into installer (build is on) or boot from local drive (build is off). This template is used to load the GRUB2 configuration file or chainload from local EFI ESP for systems which booted from network.
```
"This template tries to load first different GRUB2 configuration file from local ESP, then to chainload EFI binaries from ESP."

jloeser commented on app/views/unattended/provisioning_templates/snippet/pxegrub2_chainload.erb:
```diff
@@ -67,12 +53,7 @@ echo
-  fi
-
-  menuentry 'Loading GRUB2 config from ESP' --id grub_config {
+menuentry 'Loading GRUB2 from ESP' --id local {
```
"Loading system from ESP" would be better here because it first tries to load GRUB2 configuration files, then EFI binaries.

goarsna committed:
Fixes #37653 - Always load local disk's GRUB2 configuration

Load the local disk's GRUB2 configuration regardless of weather
SecureBoot is enabled or not. This standardizes the boot process
under UEFI.

To support SecureBoot for arbitrary operating systems, support for
distribution vendor specific boot files is added separately.

The following existing approaches for local boot don't work (anymore).

Using `chainloader` command:
    - Not supported according to GRUB2 manual [1] if SecureBoot is
      enabled (even if supported by some distribution vendors).
    - Chainloading with disabled SecureBoot requires patched
      `connectefi` command on some platforms which is currently only
      supported by EL GRUB2 binaries.

Using `exit 1` to boot from next boot device by firmware:
    - Tests showed that this behavior is not deterministic across
      different distribution vendor specific boot files.
    - Additional effort would be required to ensure the correct boot
      order.
    - This was introduced with commit b6b32045 for enabled SecureBoot
      only and was already fixed with commit aca4023ee.

For Windows we still use chainloading as there is no local `grub.cfg`.
The default EL GRUB2 which is used in this case supports the
`connectefi` command. SecureBoot verification is done by the Microsoft
certificate in the db.

[1]: https://www.gnu.org/software/grub/manual/grub/html_node/UEFI-secure-boot-and-shim.html#UEFI-secure-boot-and-shim

jloeser commented:
Merged PR https://github.com/theforeman/foreman/pull/10680 does only work in combination with this PR.

goarsna commented:
Rebased to head of `theforeman:develop`

jloeser commented:
LGTM.

sbernhard commented:
@stejskalleos Thouhgs on this PR?

stejskalleos commented:
> @stejskalleos Thouhgs on this PR?

I'm looking at it, got some issues with my local dev setup, so I can't retest it right now. It's on my list.

stejskalleos reviewed (approved):
🍏 LGTM

Provisioned Fedora 42 booted without any issues, changes make sense, let's get this in.
